src/hypertable_insert.c

```diff
@@ -0,0 +1,123 @@
+#include <postgres.h>
+#include <nodes/execnodes.h>
+#include <nodes/extensible.h>
+#include <nodes/makefuncs.h>
+#include <nodes/nodeFuncs.h>
+#include <executor/nodeModifyTable.h>
+#include <utils/rel.h>
+#include <catalog/pg_type.h>
+
+#include "hypertable_insert.h"
+#include "chunk_dispatch_info.h"
+#include "chunk_dispatch_state.h"
+
+/*
+ * HypertableInsert (with corresponding executor node) is a plan node that
+ * implements INSERTs for hypertables. It is mostly a wrapper around the
+ * ModifyTable plan node that simply calls the wrapped ModifyTable plan without
+ * doing much else, apart from some initial state setup.
+ *
+ * The wrapping is needed to setup state in the execution phase, and give access
+ * to the ModifyTableState node to sub-plan states in the PlanState tree. For
+ * instance, the ChunkDispatchState node needs to set the arbiter index list in
+ * the ModifyTableState node whenever it inserts into a new chunk.
+ */
+
+
+static void
+hypertable_insert_begin(CustomScanState *node, EState *estate, int eflags)
+{
+	HypertableInsertState *state = (HypertableInsertState *) node;
+	PlanState  *ps = ExecInitNode(&state->mt->plan, estate, eflags);
+
+	node->custom_ps = list_make1(ps);
+
+	if (IsA(ps, ModifyTableState))
+	{
+		ModifyTableState *mtstate = (ModifyTableState *) ps;
+		int			i;
+
+		/*
+		 * Find all ChunkDispatchState subnodes and set their parent
+		 * ModifyTableState node
+		 */
+		for (i = 0; i < mtstate->mt_nplans; i++)
+		{
+			if (IsA(mtstate->mt_plans[i], CustomScanState))
```

Comment:
I am not worried about performance. But wouldn't a non-CustomScanState not be correctly handled here? Shouldn't we have an else with an error then?